### PR TITLE
feat: implement RSLAD-style robust distillation in compression phase …

### LINKED ISSUE #47: [Feature]: Implement RSLAD-style robust distillation in compression phase
## Problem Statement

The compression phase (`CompressionPhase` in `nightmarenet/training/phases.py`) applies magnitude pruning but does NOT perform the "robust distillation" step described in the README:

> **Compress** - Preserve robustness, shed parameters. Adversarial robust distillation (RSLAD-style) + magnitude pruning.

The actual implementation only does pruning + optional fine-tuning:

```python
class CompressionPhase:
    def run(self, dataloader, optimizer):
        # 1. Apply magnitude pruning
        pruner = MagnitudePruner(pruning_ratio=self.pruning_ratio)
        stats = pruner.apply(self.model)
        # 2. Optional fine-tune after prune
        if self.finetune_after_prune:
            # ... standard fine-tuning loop ...
```

There is no knowledge distillation from the pre-pruned model to the pruned model. The README's claim of "RSLAD-style adversarial robust distillation" is unimplemented. This is significant because:

- Standard pruning followed by fine-tuning does NOT preserve adversarial robustness (known finding from Ye et al. 2019)
- The distillation step is what makes the compression phase part of the robustness mechanism (not just a size reduction)
- The paper draft references this but the code doesn't implement it

**Relevant files:**
- `nightmarenet/training/phases.py` (CompressionPhase.run)
- `nightmarenet/compression/pruning.py` (MagnitudePruner only)

## Proposed Solution

Add a distillation step to CompressionPhase:

1. Before pruning, snapshot the current model as the "teacher"
2. After pruning, run a distillation loop where the pruned "student" learns from the teacher's logits
3. Use KL divergence loss between teacher and student outputs (following RSLAD: use adversarial examples as distillation data, not clean data)
4. Config: `compression.distillation: true`, `compression.distillation_temperature: 4.0`, `compression.distillation_alpha: 0.7`

**Files affected:**
- `nightmarenet/compression/pruning.py` or new `nightmarenet/compression/distillation.py`
- `nightmarenet/training/phases.py` (CompressionPhase)
- `configs/default.yaml` (new config keys)

**Dependencies:** None new (KL divergence is in torch.nn.functional).

## Acceptance Criteria

- [ ] CompressionPhase performs KL distillation from pre-pruned teacher to pruned student
- [ ] Distillation uses adversarially-perturbed inputs (not clean data) per RSLAD
- [ ] Config keys added: `compression.distillation`, `compression.distillation_temperature`, `compression.distillation_alpha`
- [ ] Distillation is optional (backward compatible when disabled)
- [ ] Test: verify distillation reduces KL divergence between teacher and student outputs
- [ ] Documentation: update README compress phase description if behavior changes

## Scope / Difficulty Estimate

Medium-high. ~4-5 hours. Requires understanding the distillation literature and implementing a training loop within the compression phase.

### PR DESCRIPTION
**Pull request by RitikaSen27**
Closes #47

## Changes
- Added `nightmarenet/compression/distillation.py` with RSLAD-style distillation logic
- `CompressionPhase.run` now snapshots the pre-pruned model as a frozen teacher before pruning
- After pruning, the student (pruned model) is trained via KL divergence loss against teacher outputs on adversarially-perturbed inputs (FGSM)
- Added config keys: `distillation`, `distillation_temperature`, `distillation_alpha`, `distillation_epochs` to `configs/default.yaml`
- Distillation is opt-in (backward compatible when `distillation: false`)

## Testing
- All 492 existing tests pass

## Summary by CodeRabbit

* **New Features**
  * Added optional adversarial knowledge distillation during model compression.
  * Added configurable distillation settings, including temperature, weighting, and training epochs.
  * Compression results now include distillation metrics when enabled.
  * Updated pruning configuration to use magnitude-based pruning.

* **Bug Fixes**
  * Added safeguards to skip invalid training updates when loss values are not finite.

### CODE REVIEW

**Review by Adit-Jain-srm — CHANGES_REQUESTED** [review]
Please read and follow contributing guidelines, resolve CI / tests and lint checks error, resolve code rabbit comments, fulfill pre-submission checklist @RitikaSen27

**Comment by RitikaSen27** [comment]
Okay,I'll get this fixed by tonight
